ledger/complete/mtrie/trie/trie.go

```diff
@@ -324,13 +363,39 @@ func update(
 				// the case where the recursion stops: only one path to update
 				if len(paths) == 1 {
 					if !parentNode.Payload().Equals(&payloads[i]) {
-						return node.NewLeaf(paths[i], payloads[i].DeepCopy(), nodeHeight)
+						n = node.NewLeaf(paths[i], payloads[i].DeepCopy(), nodeHeight)
+
+						allocatedRegCountDelta = 0
+						if payloads[i].IsEmpty() {
+							// Old payload is not empty while new payload is empty.
+							// Allocated register will be unallocated.
+							allocatedRegCountDelta = -1
+						} else if parentNode.Payload().IsEmpty() {
+							// Old payload is empty while new payload is not empty.
+							// Unallocated register will be allocated.
+							allocatedRegCountDelta = 1
+						}
+
+						oldPayloadSize := parentNode.Payload().Size()
+						newPayloadSize := payloads[i].Size()
+						allocatedRegSizeDelta = int64(newPayloadSize - oldPayloadSize)
+
+						return n, allocatedRegCountDelta, allocatedRegSizeDelta, nodeHeight
 					}
 					// avoid creating a new node when the same payload is written
-					return parentNode
+					return parentNode, 0, 0, nodeHeight
 				}
 				// the case where the recursion carries on: len(paths)>1
 				found = true
+
+				if !parentNode.Payload().IsEmpty() {
+					// Allocated register will be updated or unallocated at lower height.
+					allocatedRegCountDelta--
+				}
+
```

Comment:
Ideally, we could extract these lines into a sub-function to restrict the function here to the core logic. you could directly assign the function's return values to `allocatedRegCountDelta` and `allocatedRegSizeDelta`. As this is the first modifying operation in this code path

Reply:
Good point! Updated in 4bcc1c815d300bdee1af5b5429f7b19fb4394928.